test/shared/src/main/scala/zio/test/MutableRunnableSpec.scala

```diff
@@ -0,0 +1,144 @@
+package zio.test
+
+import zio.clock.Clock
+import zio.duration._
+import zio.test.environment.TestEnvironment
+import zio.{Chunk, Has, URIO, ZIO, ZLayer}
+
+import scala.util.control.NoStackTrace
+
+/**
+ * Syntax for writing test like
+ * {{{
+ * object MySpec extends MutableRunnableSpec(layer) {
+ *   suite("foo") {
+ *     testM("name") {
+ *     } @@ ignore
+ *
+ *     test("name 2")
+ *   }
+ *   suite("another suite") {
+ *     test("name 3")
+ *   }
+ * }
+ * }}}
+ */
+class MutableRunnableSpec[R <: Has[_]](layer: ZLayer[TestEnvironment, Throwable, R])
+    extends RunnableSpec[TestEnvironment, Any] {
+  self =>
+
+  type ZS = ZSpec[R, Any]
+
+  class InAnotherTestException(`type`: String, label: String)
+      extends Exception(s"${`type`} `${label}` is in another test")
+      with NoStackTrace
+
+  sealed trait SpecBuilder {
+    def toSpec: ZS
+    def label: String
+  }
+
+  case class SuiteBuilder(label: String) extends SpecBuilder {
+
+    private[test] var nested: Chunk[SpecBuilder]                   = Chunk.empty
+    private var aspects: Chunk[TestAspect[R, R, Failure, Failure]] = Chunk.empty
+
+    /**
+     * Syntax for adding aspects.
+     * {{{
+     * test("foo") { assert(42, equalTo(42)) } @@ ignore
+     * }}}
+     */
+    final def @@(
+      aspect: TestAspect[R, R, Failure, Failure]
+    ): SpecBuilder = {
+      aspects = aspects :+ aspect
+      this
+    }
+
+    def toSpec: ZS =
+      aspects.foldLeft(
+        zio.test.suite(label)(
+          nested.map(_.toSpec): _*
+        )
+      )((spec, aspect) => spec @@ aspect)
+  }
+
+  case class TestBuilder(label: String, var toSpec: ZS) extends SpecBuilder {
+
+    /**
+     * Syntax for adding aspects.
+     * {{{
+     * test("foo") { assert(42, equalTo(42)) } @@ ignore
+     * }}}
+     */
+    final def @@(
+      aspect: TestAspect[R, R, Failure, Failure]
+    ): TestBuilder = {
+      toSpec = toSpec @@ aspect
+      this
+    }
+  }
+
+  // init SpecBuilder for this test class
+  var stack: List[SuiteBuilder] = SuiteBuilder(self.getClass.getSimpleName.stripSuffix("$")) :: Nil
+  var testRunning               = false
+
+  /**
+   * Builds a suite containing a number of other specs.
+   */
+  def suite(label: String)(specs: => SpecBuilder): SuiteBuilder = {
```

Comment:
I'd make all the methods in here `final` unless there is a good reason not to do so.

Reply:
Fixed